android/expoview/src/main/java/host/exp/exponent/kernel/DevMenuManager.kt

```diff
@@ -0,0 +1,317 @@
+package host.exp.exponent.kernel
+
+import android.content.Context
+import android.content.pm.ActivityInfo
+import android.hardware.SensorManager
+import android.os.Bundle
+import android.view.View
+import android.view.ViewGroup
+import android.widget.FrameLayout
+import com.facebook.react.ReactRootView
+import com.facebook.react.bridge.Arguments
+import com.facebook.react.bridge.ReadableMap
+import com.facebook.react.bridge.UiThreadUtil
+import com.facebook.react.bridge.WritableMap
+import host.exp.exponent.utils.ShakeDetector
+import host.exp.exponent.Constants
+import versioned.host.exp.exponent.modules.internal.DevMenuModule
+import host.exp.exponent.di.NativeModuleDepsProvider
+import host.exp.exponent.experience.ExperienceActivity
+import versioned.host.exp.exponent.ReactUnthemedRootView
+import java.util.*
+import javax.inject.Inject
+import host.exp.exponent.modules.ExponentKernelModule
+import host.exp.exponent.storage.ExponentSharedPreferences
+
+
+private const val DEV_MENU_JS_MODULE_NAME = "HomeMenu"
+
+/**
+ * DevMenuManager is like a singleton that manages the dev menu in the whole application
+ * and delegates calls from [ExponentKernelModule] to the specific [DevMenuModule]
+ * that is linked with a react context for which the dev menu is going to be rendered.
+ * Its instance can be injected as a dependency of other classes by [NativeModuleDepsProvider]
+ */
+class DevMenuManager {
+  private var shakeDetector: ShakeDetector? = null
+  private var reactRootView: ReactRootView? = null
+  private var orientationBeforeShowingDevMenu: Int = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
+  private val devMenuModulesRegistry = WeakHashMap<ExperienceActivity, DevMenuModuleInterface>()
+
+  @Inject
+  internal val kernel: Kernel? = null
+
+  @Inject
+  internal val exponentSharedPreferences: ExponentSharedPreferences? = null
+
+  init {
+    NativeModuleDepsProvider.getInstance().inject(DevMenuManager::class.java, this)
+  }
+
+  //region publics
+
+  /**
+   * Links given [DevMenuModule] with given [ExperienceActivity]. [DevMenuManager] needs to know this to
+   * get appropriate data or pass requests down to the correct [DevMenuModule] that handles
+   * all these stuff for a specific experience (DevMenuManager only delegates those calls).
+   */
+  fun registerDevMenuModuleForActivity(devMenuModule: DevMenuModuleInterface, activity: ExperienceActivity) {
+    // Start shake detector once the first DevMenuModule registers in the manager.
+    maybeStartDetectingShakes(activity.applicationContext)
+    devMenuModulesRegistry[activity] = devMenuModule
+  }
+
+  /**
+   * Shows dev menu in given experience activity. Ensures it never happens in standalone apps and is run on the UI thread.
+   */
+  fun showInActivity(activity: ExperienceActivity) {
+    if (Constants.isStandaloneApp()) {
```

Comment:
That might be harder to do because it's not RN module. It's provided by `NativeModuleDepsProvider` so it's injected into some other classes that can use and call these methods on their own. We would need to check whether the manager is not null in those places.